File: backend/app/routers/payments.py
from fastapi import APIRouter, Depends
from app.dependencies import verify_token

router = APIRouter(prefix="/api/payments", tags=["payments"])


@router.get("/")
async def list_payments(auth: dict = Depends(verify_token)):
    from app.database import get_supabase
    supabase = get_supabase()
    profiles = supabase.table("profiles").select("role").eq("id", auth["user_id"]).execute()
    if not profiles.data:
        return {"success": True, "data": []}
    role = profiles.data[0].get("role")

    if role == "student":
        students = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
        if not students.data:
            return {"success": True, "data": []}
        contracts = supabase.table("contracts").select("id").eq("student_id", students.data[0]["id"]).execute()
    elif role == "company":
        companies = supabase.table("company_profiles").select("id").eq("user_id", auth["user_id"]).execute()
        if not companies.data:
            return {"success": True, "data": []}
        contracts = supabase.table("contracts").select("id").eq("company_id", companies.data[0]["id"]).execute()
    else:
        result = supabase.table("payments").select("*").order("due_date", desc=True).execute()
        return {"success": True, "data": result.data}

    ids = [c["id"] for c in contracts.data] if contracts.data else []
    if not ids:
        return {"success": True, "data": []}
    result = supabase.table("payments").select("*").in_("contract_id", ids).order("due_date", desc=True).execute()
    return {"success": True, "data": result.data}

File: backend/src/routes/payments.ts
import { Router, Request, Response } from "express";
import { verifyToken } from "../middleware/auth";
import { supabase } from "../database";

const router = Router();

router.get("/", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data: profiles } = await supabase
      .from("profiles")
      .select("role")
      .eq("id", req.user!.user_id);
    if (!profiles || profiles.length === 0) {
      return res.json({ success: true, data: [] });
    }
    const role = profiles[0].role;
    let contractIds: string[] = [];
    if (role === "student") {
      const { data: students } = await supabase
        .from("student_profiles")
        .select("id")
        .eq("user_id", req.user!.user_id);
      if (!students || students.length === 0) return res.json({ success: true, data: [] });
      const { data: contracts } = await supabase
        .from("contracts")
        .select("id")
        .eq("student_id", students[0].id);
      contractIds = contracts ? contracts.map((c) => c.id) : [];
    } else if (role === "company") {
      const { data: companies } = await supabase
        .from("company_profiles")
        .select("id")
        .eq("user_id", req.user!.user_id);
      if (!companies || companies.length === 0) return res.json({ success: true, data: [] });
      const { data: contracts } = await supabase
        .from("contracts")
        .select("id")
        .eq("company_id", companies[0].id);
      contractIds = contracts ? contracts.map((c) => c.id) : [];
    } else {
      const { data } = await supabase
        .from("payments")
        .select("*")
        .order("due_date", { ascending: false });
      return res.json({ success: true, data: data || [] });
    }
    if (contractIds.length === 0) return res.json({ success: true, data: [] });
    const { data } = await supabase
      .from("payments")
      .select("*")
      .in("contract_id", contractIds)
      .order("due_date", { ascending: false });
    return res.json({ success: true, data: data || [] });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

export default router;
